Monika-After-Story/MonikaModDev
# New D25

# New D25:

# Additions:
## Lots of new functions:
### `mas_getPlayerAge()`:
- Gets the player's age (based off player birthday)
### `mas_canShowRisque()`:
- Returns T/F if we can show risque things (See docstring for condition details)
### `mas_wasFirstValueIn()`:
- Checks history archives to see if the first year a value (you provide) at the key (you provide) matches the year (you provide)
### `mas_getFirstYearOfValue()`:
- Checks history to get the first year the value (you provide) at the key (you provide) is stored
### `mas_verifyHistAll_k()`:
- Checks if the value (you provide) at the key (you provide) is present in history at all
### `mas_verifyHistLastYear_k()`:
- Checks if the value (you provide) at the key (you provide) is present in history last year
### `mas_lastSeenLastYear()`:
- Checks if the eventlabel provided was last seen last year
### `mas_addClothesToHolidayMapRange()`:
- Ranged variant of `mas_addClothesToHolidayMap()`, which will add the specified clothes to all the days within the start/end dates provided
### `mas_isFirstSeshPast()`:
- Checks if the first session is past the date you provide
### `mas_d25ShowVisuals()`:
- Shows D25 visuals
### `mas_d25HideVisuals()`:
- Hides D25 visuals
### `mas_isD25Pre()`:
- Checks if we're within the d25 season, but before d25
### `mas_isD25GiftHold()`:
- Checks if we're within the range we should keep the gifts under the tree
### `mas_d25ReactToGifts()`:
- Sets up reactions for gifts under the tree
### `mas_d25SilentReactToGifts()`:
- Silently does the gift reactions (no dialogue) for the gifts under the tree
### `mas_d25_utils.shouldUseD25ReactToGifts()`
- Checks if we should use the `mas_d25_utils.react_to_gifts()` function to register gifts
### `mas_d25_utils.react_to_gifts()`:
- Registers gifts under the d25 protocol
### `mas_island_event.getBackground()`:
- Gets the background for the islands (Winter forces snow islands, the rest uses the ones built into the weather object)
### `mas_hasUnlockedClothesWithExprop()`:
- Checks if we have unlocked clothes with the exprop (and optionally, value) provided
### `mas_getGiftedDates()`:
- Gets the dates that the gift provided was gifted on (as a list) (pass in gift reaction label)
### `mas_lastGiftedInYear()`:
- Checks if a gift was last gifted in the year provided (pass in gift reaction label)

## New class: `MASPoem`
- See `zz_poems.rpy` for details and functions included there

## Other Additions:
- New deco
- New outfits
- New topics
- New stories
- New songs
- New Finalfarewell poem
- New Finalfarewell flow/other ways to get into the flow for the FF
- Spritechecker can now check for `extend` lines and their expression validity.
- `mas_clothes_change` label accepts more params. (exp, restore_zoom, unlock)
  - exp: exp to show on return
  - restore_zoom: whether or not we restore zoom on return
  - unlock: whether or not we wish to unlock the clothes we're changing into directly
- `mas_songs.getVolume()` gets `actual_volume` now
- `addReaction()` accepts an `exclude_on` parameter, which takes a list of exclusion keys
- `mas_showpoem` label, which handles both `MASPoem`s and normal `Poem`s
- Also added a topic to see Monika's old poems again
- All weather is actually properly unlocked now
- Promisering naming is no longer cursed and actually makes sense
- Islands have all their bgs + dlg for them
- Undo action rules can no longer be adjusted to have NoneTypes stored in them
- `mas_isplayer_bday()` now takes in an extra optional parameter, `use_date_year`, which will use the year from `_date`. This defaults `False`.

## Testing:
- Make sure normal flow works, no crashes, no out of place dlg (wrong history mentioned)
- Make sure nts isn't triggered if you start a new persist post d25
- Make sure all new clothes get unlocked
- Make sure gift reactions work properly for the d25 protocol
- Verify both ways of exiting the d25 season work properly
